### PR TITLE
Alerting: make alert rule routine evaluation control be thread-safe

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
The ngalert scheduler has an internal structure (so-called `alertRuleInfo`) that controls the rule evaluation loop. The control structure contains three fields: 
- `evalCh` - a channel that is used to tell the rule routine to evaluate a rule
- `stopCh` - a channel that tells the rule routine to exit
- `version` - a version of the rule at the point of time.

When the scheduler runs a synchronization loop, which fetches the rules from the database, and decides whether an evaluation needs to be run or stopped for a rule, it requests the control structure for the rule from a registry, which is just a thread-safe map with some nuances. The nuance is that the registry returns a copy of the control structure. The reason for that seems to be to ensure that the evaluation that is scheduled during the previous sync would not be changed if the version of the rule is changed during the current sync cycle (@papagian please correct me if I am wrong).

This PR moves the version from the `alertRuleInfo` to an intermediate structure `readyToRunItem` that is disposed after the evaluation context was submitted to the evaluation channel. Also, it updates the registry to return a pointer instead of the copy.  The reason for that is to make sure that all goroutines use the same instance of the controlling structure (see below for motivation)

Also, the control structure does not enforce consistency between stopCh` and `evalCh`, which can lead to the races when the control structure is accessed from different threads.

<details><summary>Example</summary

Note: this does not reflect the current situation. It reflects just a potential problem when eval and stop channels are not synchronized.

image

In the diagram above the second evaluation will hang indefinitely because a goroutine that pushes the data in the eval channel does not know that it was stopped. Also, if I understand the mechanics correctly, that goroutine will hold the reference to the channel and therefore it will not be collected by GC. 
</details>

This PR removes the channel `stopCh` from `alertRuleInfo` and replaces it with cancellable context. 
Also, it adds a function Eval that safely calls sends to `evalCh` and exists if the context is canceled.

**Which issue(s) this PR fixes**:
Related: https://github.com/grafana/grafana/pull/39975

**Special notes for your reviewer**:
This particular PR does not resolve any particular bug or issue. It should be reviewed in the context of changes coming as part of PR  https://github.com/grafana/grafana/pull/39975 that introduces another channel to the control struct + ways to control the evaluation outside of they synchronization routine

### CODE REVIEW

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:378]: I'm not sure I understand this change when the version is also in `ruleInfo`?

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:586]: Don't we know that its a new struct as the version in the `alertRuleInfo` is `0`?

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:641]: I don't understand the need for `*atomic.Bool` when we also lock `mtx`? Can't this just be a `bool`?

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:345]: We are changing this from a `alertRuleInfo` to a `*alertRuleInfo` because it now holds a mutex right?

**Review by grobinson-grafana — COMMENTED** [review]
I have left some comments as I am not sure about this change. It's not the intention of the change but I am not sure about `alertRuleInfo` being mutable and containing both a mutex and a thread-safe bool?

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:649]: In the past I've just done something like:

```suggestion
func (a *alertRuleInfo) Stop() bool {
     a.mtx.Lock()
     defer a.mtx.Unlock()
     if !a.stopped {
          close(a.stopCh)
          a.stopped = true
     }
     return a.stopped
}
```

Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:649]: In the past I've just done something like:

```suggestion
func (a *alertRuleInfo) Stop() bool {
     a.mtx.Lock()
     defer a.mtx.Unlock()
     if !a.stopped {
          close(a.stopCh)
          a.stopped = true
     }
     return a.stopped
}
```

Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:649]: There is a pattern called "double-checked locking", which implies that the expression is evaluated before and after acquiring a lock to avoid\reduce congestion in the code wrapped by the mutex. I think in this case and this particular design of the rule evaluation routine, this approach is necessary. 
The code you suggest (I assume the same pattern you suggest using for Eval (and Update, not in this PR) as well) has a problem:
when one goroutine calls Stop, it can end up waiting for ruleRoutine to fetch data from the `stopCh`. The evaluation loop does not guarantee how fast it will read that channel because when it handles evaluation it does some pretty time-consuming operations (calls external services, retries). Therefore, the mutex can be locked for a substantial time.  
When during that time some other goroutine will try to interact with the control structure alertRuleInfo (Stop - can come from an API (other PR) as well as scheduler's sync), Update - can come from multiple concurrent requests, Eval - can come from scheduler's sync) it will be blocked on the mutex just to check whether it is closed or not. but that information is already known (it is closed at that time).

Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:649]: The reason I used atomic boolean was that there is a guess that using a regular bool can make race detector complain about it.

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:641]: The reason I used atomic boolean was that there is a guess that using a regular bool can make race detector complain about it.

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:378]: I moved the version from the `alertRuleInfo` to `readyToRunItem` because the control structure is not copied any more.

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:586]: version was removed from the structure. This code replaces exists + getOrCreate here 
https://github.com/grafana/grafana/pull/41220/files#diff-1fe393768892c6c76a43e1ac0ea9ca69ec72d2124d6d80abbede394dd49077aeL350-L351

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:345]: Correct

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:649]: Oh, I just realized that I (as well as in your suggestion) do not send data to stopCh (before I sent a message as we do currently). So, yes, your suggestion does make sense.

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go]: I applied your suggestion and there is no need in the atomic bool anymore. So I replaced it with just bool

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:660]: Here I don't think we need to acquire the mutex, we can just check if `a.stopCh` is closed. `a.stopped` still needs to be set to true to make sure we don't close a closed channel on successive calls to `Stop()`.

```suggestion
	select {
	    case <-a.stopCh:
	        return false
	    default:
	}
```

Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:660]: Here I don't think we need to acquire the mutex, we can just check if `a.stopCh` is closed. `a.stopped` still needs to be set to true to make sure we don't close a closed channel on successive calls to `Stop()`.

```suggestion
	select {
	    case <-a.stopCh:
	        return false
	    default:
	}
```

Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:660]: Here I don't think we need to acquire the mutex, we can just check if `a.stopCh` is closed. `a.stopped` still needs to be set to true to make sure we don't close a closed channel on successive calls to `Stop()`.

```suggestion
	select {
	    case <-a.stopCh:
	        return false
	    default:
	}
```

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:659]: Here I don't think we need to acquire the mutex, we can just check if `a.stopCh` is closed. However `a.stopped` still needs to be set to true to make sure we don't close a closed channel on successive calls to `Stop()`.

```suggestion
	select {
	    case <-a.stopCh:
	        return false
	    default:
	}
```

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:665]: `eval.Ch` is unbuffered and so this operation will block until `ruleRoutine` has completed it's most recent evaluation of the alert rule. That's often not an issue except we have also acquired the mutex at the start of the function?

Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:665]: I think the mutex here is still useful. Yes, it will block the other "commands". The reason for having it is that it guards against the following situation
image

It is possible because of the select that picks a channel that has data in a pseudo-random way (https://golang.org/ref/spec#Select_statements). I am not 100% sure whether it works the same way in the case of a closed channel, though.

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:659]: please see https://github.com/grafana/grafana/pull/41220#discussion_r744816199

**Comment by yuri-tceretian** [comment]
Discussed with @gerobinson in the slack and we decided that it would make sense to get rid of `stopCh` at all and use context cancellation. https://github.com/grafana/grafana/pull/41220/commits/01f02ad86e82538d1db1b858bab2b0518bf06f96 implements that.

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:655]: This will panic when `a.evalCh` is closed, need to make sure that `Eval` does not run if `a.evalCh` is closed. I think it's best if `Eval` is responsible for closing `a.evalCh` then we don't need to think about writing into a closed channel, however, closing a channel should be done at most once so we will also need a marker.

Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:655]: You're correct about panic, but we do not close the channel. This select makes the sender to stop sending data to a channel when the context is already cancelled and there is no consumer on other side. Due to nature of the scheduler this can happen when rule evaluation takes more time than it's evaluation interval which can cause evaluations to be piled up. When rule is deleted, we remove the rule from registry and stop the context. That causes those piled up evaluation to be cancelled immediatelly, and therefore, release the pointer to the channel so it could be collected by GC.

File: pkg/services/ngalert/schedule/schedule.go
Comment (davidmparrott) [inline pkg/services/ngalert/schedule/schedule.go:644]: Why is Stop exported here? Only remaining question I have beyond what George has already asked.

**Review by davidmparrott — APPROVED** [review]

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:644]: No reason. I will change

**Comment by gotjosh** [comment]
I'd hold off on merging this until @gerobinson co-approves as he is one of the main reviewers and he hasn't addressed your last comment.

**Review by gotjosh — APPROVED** [review]
LGTM

I'm a bit concerned about making changes in this area - mostly because I don't know enough about it to map all the changes in my head.

I think that the removal of the evaluation channel in favour of making the rule routine being context-aware is _an_ improvement, but my only doubt is whenever this is worth the _actual_ change. Any unexpected behaviour might be hard to track and will directly affect customers.

That being said, I'll leave the final decision to the author.

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:534]: We shouldn't delete this code but since the channel is never closed we will never enter this conditional block.

**Review by grobinson-grafana — APPROVED** [review]

**Comment by gerobinson** [comment]
I had another look at this and couldn't see a cause for concern.

I would like us to think about doing a wider refactoring of the scheduler sooner rather than later as it's much more complicated, and much harder to understand, than it needs to be. The first change I think we should make is separating scheduling from the actual evaluation of an alert rule which at present both happen together. Second I think we should change how we schedule evaluation goroutines and the evaluation of alert rules via `evalCh`.

**Pull request merged**
